icu4j/main/classes/core/src/com/ibm/icu/util/MeasureUnit.java

```diff
@@ -552,7 +545,21 @@ public boolean equals(Object rhs) {
             return false;
         }
         MeasureUnit c = (MeasureUnit) rhs;
-        return type.equals(c.type) && subType.equals(c.subType);
+        if (type != null && subType != null && type.equals(c.type) && subType.equals(c.subType) ) {
+            return true;
+        }
+
+        if (this.measureUnitImpl ==  ((MeasureUnit) rhs).measureUnitImpl) {
+            return true;
+        }
+
+        if (this.measureUnitImpl == null || ((MeasureUnit) rhs).measureUnitImpl == null) {
+            // If both of them were null, the previous if statement would be true.
+            // And the return value would be true.
+            return false;
+        }
+
+        return this.getIdentifier().equals(((MeasureUnit) rhs).getIdentifier());
```

Comment:
Does it work if you use this line as your equality check, instead of lines 548-560?

Reply:
Yes, you are right.

or I can compare change 562 to `return this.measureUnitImpl.equals(rhs.measureUnitImpl))`

Reply:
we can just leave `562` for now and remove the others.